neo/Cryptography/Base58.cs

```diff
@@ -11,45 +11,50 @@ public static class Base58
 
         public static byte[] Decode(string input)
         {
-            BigInteger bi = BigInteger.Zero;
-            for (int i = input.Length - 1; i >= 0; i--)
+            // Decode Base58 string to BigInteger 
+            var bi = BigInteger.Zero;
+            for (int i = 0; i < input.Length; i++)
             {
-                int index = Alphabet.IndexOf(input[i]);
-                if (index == -1)
-                    throw new FormatException();
-                bi += index * BigInteger.Pow(58, input.Length - 1 - i);
+                int digit = Alphabet.IndexOf(input[i]);
+                if (digit < 0)
+                    throw new FormatException(string.Format("Invalid Base58 character `{0}` at position {1}", input[i], i));
+                bi = bi * 58 + digit;
             }
-            byte[] bytes = bi.ToByteArray();
-            Array.Reverse(bytes);
-            bool stripSignByte = bytes.Length > 1 && bytes[0] == 0 && bytes[1] >= 0x80;
-            int leadingZeros = 0;
-            for (int i = 0; i < input.Length && input[i] == Alphabet[0]; i++)
-            {
-                leadingZeros++;
-            }
-            byte[] tmp = new byte[bytes.Length - (stripSignByte ? 1 : 0) + leadingZeros];
-            Array.Copy(bytes, stripSignByte ? 1 : 0, tmp, leadingZeros, tmp.Length - leadingZeros);
-            Array.Clear(bytes, 0, bytes.Length);
-            return tmp;
+
+            // Encode BigInteger to byte[]
+            // Leading zero bytes get encoded as leading `1` characters
+            int leadingZeroCount = input.TakeWhile(c => c == '1').Count();
+            var leadingZeros = new byte[leadingZeroCount];
+            var bytesWithoutLeadingZeros =
+                bi.ToByteArray()
+                .Reverse()// to big endian
+                .SkipWhile(b => b == 0);//strip sign byte
+            return leadingZeros.Concat(bytesWithoutLeadingZeros).ToArray();
         }
 
         public static string Encode(byte[] input)
         {
-            BigInteger value = new BigInteger(new byte[1].Concat(input).Reverse().ToArray());
-            StringBuilder sb = new StringBuilder();
-            while (value >= 58)
+            // Decode byte[] to BigInteger
+            var intData = BigInteger.Zero;
+            for (int i = 0; i < input.Length; i++)
+            {
+                intData = intData * 256 + input[i];
+            }
```

Comment:
I think there is no difference between this and the old lines:

```c#
BigInteger value = new BigInteger(new byte[1].Concat(input).Reverse().ToArray());
```

Reply:
Yes, solved